src/cdk-experimental/menu/menu-stack.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Subject, Observable} from 'rxjs';
+
+/** Events to emit as specified by the caller once the MenuStack is empty. */
+export type FocusNext = 'nextItem' | 'previousItem' | 'currentItem' | undefined;
+
+/**
+ * Interface for the elements tracked in the MenuStack.
+ */
+export interface MenuStackItem {
+  /** A reference to the previous Menus MenuStack instance. */
+  _menuStack: MenuStack;
+}
+
+/**
+ * MenuStack allows subscribers to listen for close events (when a MenuStackItem is popped off
+ * of the stack) in order to perform closing actions. Upon the MenuStack being empty it emits
+ * from the `empty` observable specifying the next focus action which the listener should perform
+ * as requested by the closer.
+ */
+export class MenuStack {
+  /** All MenuStackItems tracked by this MenuStack. */
+  private readonly _elements: MenuStackItem[] = [];
+
+  /** Emits the element which was popped off of the stack when requested by a closer. */
+  private readonly _close: Subject<MenuStackItem> = new Subject();
+
+  /** Emits once the MenuStack has become empty after popping off elements. */
+  private readonly _empty: Subject<FocusNext> = new Subject();
+
+  /** Observable which emits the MenuStackItem which has been requested to close. */
+  close(): Observable<MenuStackItem> {
```

Comment:
We have an internal guideline to use typing to restrict the type when exposing a subject as an observable, instead of explicitly calling asObservable(). @jelbourn , is this also true in the CDK codebase, or do you prefer to call asObservable()?

Reply:
Yeah, we could do that here; that is probably better

Reply:
Was already merged in before I had a chance to address, will include fix in a follow on pr.